File: ios/RecipeClipper/App/FeatureFlags.swift
import Foundation
import Observation

/// Every feature flag (#87), one case per entry in `shared/flags.json`, which both apps read.
/// `FeatureFlagsTests` fails if the two lists differ, or if a flag is no longer referenced
/// anywhere: retiring a flag deletes it here, in flags.json and its branches in one PR.
/// Android's twin is `Flag` in `data/flags/FeatureFlags.kt`, with the same keys.
enum Flag: String, CaseIterable {
    /// The meal plan (#47, #49–#51): the bottom tabs and the recipe screen's plan actions.
    case mealPlan
    /// Chef mode (#100): the Settings switch for short steps written on the device.
    case chefMode
    /// Ingredient amounts inside steps (#101): the Settings switch and what it shows.
    case amountsInSteps
    /// The free tier (#107): 20 recipes, and the one-time unlock for unlimited ones.
    case freeTier
    /// A recipe picked from the page's text by the on-device model when the page has no recipe data (#103).
    case llmExtraction
    /// Typed decisions by the on-device model: count brackets, close pantry names, aisles (#104).
    case aiDecisions
    /// "I made this" (#116): your photos and notes on a recipe, and the Recently cooked sort.
    case cookedPhotos
}

/// One flag as `shared/flags.json` declares it.
struct FlagDefinition: Equatable, Decodable {
    struct Defaults: Equatable, Decodable {
        let debug: Bool
        let release: Bool
    }

    let key: String
    let description: String
    let defaults: Defaults
    let issue: Int

    func defaultValue(isDebug: Bool) -> Bool { isDebug ? defaults.debug : defaults.release }
}

/// `shared/flags.json`, bundled beside the app like the shared tables.
enum FlagRegistry {
    private struct File: Decodable { let flags: [FlagDefinition] }

    static func parse(_ data: Data) throws -> [FlagDefinition] {
        try JSONDecoder().decode(File.self, from: data).flags
    }

    static let definitions: [FlagDefinition] = {
        guard let url = Bundle.main.url(forResource: "flags", withExtension: "json"),
              let data = try? Data(contentsOf: url),
              let flags = try? parse(data) else {
            assertionFailure("flags.json is missing from the bundle or malformed")
            return []
        }
        return flags
    }()

    static var isDebugBuild: Bool {
        #if DEBUG
        true
        #else
        false
        #endif
    }
}

/// Where the Developer settings overrides live: only the flags that differ from their default,
/// by key. `UserDefaultsFeatureFlagStore` is the real one (its own suite, never the settings');
/// `MemoryFeatureFlagStore` holds them for a run.
protocol FeatureFlagStore: AnyObject {
    var overrides: [String: Bool] { get }
    /// nil removes the override, so the flag follows its default again.
    func setOverride(_ key: String, _ value: Bool?)
    func clear()
}

final class UserDefaultsFeatureFlagStore: FeatureFlagStore {
    /// Its own suite, apart from the settings in the App Group suite, so a reset can never touch
    /// those. The share extension never reads flags, so it stays out of the App Group.
    static let suiteName = "RecipeClipperFeatureFlags"

    private let defaults: UserDefaults
    private let suite: String

    init(suiteName: String = suiteName) {
        suite = suiteName
        defaults = UserDefaults(suiteName: suiteName) ?? .standard
    }

    var overrides: [String: Bool] {
        (defaults.persistentDomain(forName: suite) ?? [:]).compactMapValues { $0 as? Bool }
    }

    func setOverride(_ key: String, _ value: Bool?) {
        if let value { defaults.set(value, forKey: key) } else { defaults.removeObject(forKey: key) }
    }

    func clear() {
        defaults.removePersistentDomain(forName: suite)
    }
}

final class MemoryFeatureFlagStore: FeatureFlagStore {
    private(set) var overrides: [String: Bool]

    init(_ overrides: [String: Bool] = [:]) { self.overrides = overrides }

    func setOverride(_ key: String, _ value: Bool?) { overrides[key] = value }
    func clear() { overrides = [:] }
}

/// Typed access to the flags: each is its build type's default from `shared/flags.json` unless
/// the store holds an override. Observable, so a view reading `isOn` (the root's tab shell)
/// redraws when Developer settings changes it, with no restart. Platform-free.
@MainActor
@Observable
final class FeatureFlags {
    let definitions: [FlagDefinition]
    @ObservationIgnored private let store: FeatureFlagStore
    @ObservationIgnored private let isDebug: Bool
    private var overrides: [String: Bool]

    init(
        store: FeatureFlagStore,
        definitions: [FlagDefinition] = FlagRegistry.definitions,
        isDebug: Bool = FlagRegistry.isDebugBuild
    ) {
        self.store = store
        self.definitions = definitions
        self.isDebug = isDebug
        overrides = store.overrides
    }

    func definition(_ flag: Flag) -> FlagDefinition? { definitions.first { $0.key == flag.rawValue } }

    func defaultValue(_ flag: Flag) -> Bool { definition(flag)?.defaultValue(isDebug: isDebug) ?? false }

    func isOn(_ flag: Flag) -> Bool { overrides[flag.rawValue] ?? defaultValue(flag) }

    func isOverridden(_ flag: Flag) -> Bool { overrides[flag.rawValue] != nil }

    /// Choosing the default removes the override, so a later change of default still applies.
    func set(_ flag: Flag, _ on: Bool) {
        store.setOverride(flag.rawValue, on == defaultValue(flag) ? nil : on)
        overrides = store.overrides
    }

    func reset() {
        store.clear()
        overrides = store.overrides
    }

    /// The store key of `unlockedOverride`; the same on Android.
    static let unlockedOverrideKey = "override.unlocked"

    /// Developer settings' "Unlocked" (#107): the unlock counts as bought, to test the unlimited
    /// library without a store. Kept in the same store under a key that is no flag's, so Reset
    /// clears it too.
    var unlockedOverride: Bool { overrides[Self.unlockedOverrideKey] == true }

    func setUnlockedOverride(_ on: Bool) {
        store.setOverride(Self.unlockedOverrideKey, on ? true : nil)
        overrides = store.overrides
    }
}

File: app/src/main/java/com/example/recipeclipper/data/flags/FeatureFlags.kt
package com.example.recipeclipper.data.flags

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.map
import org.json.JSONObject

/**
 * Every feature flag (#87), one constant per entry in `shared/flags.json`, which both apps
 * read. `FeatureFlagRegistryTest` fails if the two lists differ, or if a flag is no longer
 * referenced anywhere: retiring a flag deletes it here, in flags.json and its branches in one PR.
 */
enum class Flag(val key: String) {
    /** The meal plan (#47, #49–#51): the bottom tabs and the recipe screen's plan actions. */
    MEAL_PLAN("mealPlan"),

    /** Chef mode (#100): the Settings switch for short steps written on the device. */
    CHEF_MODE("chefMode"),

    /** Ingredient amounts inside steps (#101): the Settings switch and what it shows. */
    AMOUNTS_IN_STEPS("amountsInSteps");

    companion object {
        fun forKey(key: String): Flag? = entries.firstOrNull { it.key == key }
    }
}

/** One flag as `shared/flags.json` declares it. */
data class FlagDefinition(
    val key: String,
    val description: String,
    val debugDefault: Boolean,
    val releaseDefault: Boolean,
    val issue: Int
) {
    fun defaultFor(isDebug: Boolean) = if (isDebug) debugDefault else releaseDefault
}

/** `shared/flags.json`, read from the Java resources like the shared tables. Pure. */
object FlagRegistry {
    const val RESOURCE = "/flags.json"

    val definitions: List<FlagDefinition> by lazy {
        val stream = FlagRegistry::class.java.getResourceAsStream(RESOURCE)
            ?: error("$RESOURCE is missing from the resources")
        parse(stream.bufferedReader().use { it.readText() })
    }

    fun parse(json: String): List<FlagDefinition> {
        val flags = JSONObject(json).getJSONArray("flags")
        return (0 until flags.length()).map { i ->
            val flag = flags.getJSONObject(i)
            val defaults = flag.getJSONObject("defaults")
            FlagDefinition(
                key = flag.getString("key"),
                description = flag.getString("description"),
                debugDefault = defaults.getBoolean("debug"),
                releaseDefault = defaults.getBoolean("release"),
                issue = flag.getInt("issue")
            )
        }
    }
}

/**
 * Where the Developer settings overrides live: only the flags that differ from their default,
 * by key. [SharedPrefsFeatureFlagStore] is the real one (its own prefs file, never
 * `unit_preferences`); tests use `FakeFeatureFlagStore`.
 */
interface FeatureFlagStore {
    /** The current overrides, then every change. */
    val overrides: StateFlow<Map<String, Boolean>>

    /** [value] null removes the override, so the flag follows its default again. */
    fun setOverride(key: String, value: Boolean?)

    fun clear()
}

/** Which flags are on, at one moment. */
data class FlagValues(private val on: Set<Flag>) {
    fun isOn(flag: Flag) = flag in on

    companion object {
        /** Everything off: what a composable sees when nothing provides the real values. */
        val ALL_OFF = FlagValues(emptySet())
    }
}

/**
 * Typed access to the flags: each is its build type's default from `shared/flags.json`
 * unless the store holds an override. Platform-free, so ViewModels and tests use it as is.
 */
class FeatureFlags(
    private val store: FeatureFlagStore,
    val definitions: List<FlagDefinition>,
    private val isDebug: Boolean
) {
    private val byKey = definitions.associateBy { it.key }

    fun default(flag: Flag): Boolean = byKey[flag.key]?.defaultFor(isDebug) ?: false

    fun definition(flag: Flag): FlagDefinition? = byKey[flag.key]

    fun isOn(flag: Flag): Boolean = store.overrides.value[flag.key] ?: default(flag)

    fun isOverridden(flag: Flag): Boolean = flag.key in store.overrides.value

    val current: FlagValues get() = valuesOf(store.overrides.value)

    /** The current values, then every change. */
    val values: Flow<FlagValues> = store.overrides.map(::valuesOf).distinctUntilChanged()

    /** Choosing the default removes the override, so a later change of default still applies. */
    fun set(flag: Flag, on: Boolean) {
        store.setOverride(flag.key, if (on == default(flag)) null else on)
    }

    fun reset() = store.clear()

    private fun valuesOf(overrides: Map<String, Boolean>) =
        FlagValues(Flag.entries.filter { overrides[it.key] ?: default(it) }.toSet())
}
